tests/tests/test_mender_connect.py

```diff
@@ -176,11 +176,12 @@ def test_bogus_proto_message(
             msg = prot.encode(b"")
             ws.send(msg)
 
-            msg = ws.recv()
-            body = prot.decode(msg)
-            assert prot.props["status"] == protomsg.PROP_STATUS_ERROR
-            assert prot.protoType == 12345
-            assert prot.typ == proto_shell.MSG_TYPE_SPAWN_SHELL
+            data = ws.recv()
+            rsp = protomsg.ProtoMsg(0xFFFF)
+            rsp.decode(data)
+            assert rsp.typ == "error"
+            body = rsp.body
+            assert isinstance(body.get("err"), str) and len(body.get("err")) > 0
```

Comment:
Aren't errors expressed through the `status` field anymore? I'm a bit worried about changing basic protocol properties, mender-connect is already released, and works the way that the old test worked. Will this impact that?

Reply:
The `shell` protocol is [not affected](https://github.com/mendersoftware/mender-connect/blob/b5db81a6e8025fb3ecd453a63785c603ab1b6dca/app/daemon.go#L508). However, this test case is sending a message with an unregistered protocol header (malformed message), for which the behavior hasn't really been specified on the client side ([until now](https://github.com/mendersoftware/deviceconnect/pull/45)).

Reply:
Maybe it has not been specified in text, but it has been in code. The reason I insisted on writing this test in the first place was to have this unregistered protocol header case covered. If you send an unregistered protocol message to an existing client in the field, you will receive what is expressed in the removed code above.

So we cannot remove this. If we want to change the way that this is reported, we have to make sure that clients are prepared to deal with both error responses. But is it worth having that complexity?

What is the reason that the original error responses can not be used?

Reply:
> So we cannot remove this. If we want to change the way that this is reported, we have to make sure that clients are prepared to deal with both error responses. But is it worth having that complexity?

I don't think the UI nor `mender-cli` actually accounts for this case.

> What is the reason that the original error responses can not be used?

These error responses are used as long as the `proto` header is equal to 1 (shell). I introduced a new reserved protocol for control messages that are only intended for managing sessions, and in this protocol I added an error message for expressing internal errors (such as session expiry, no handler registered for protocol, etc.). The problem with using the shell logic is that is going to break any message that has an actual schema. Why? Because you don't know if what you get is a binary blob of error text or an encoded message (e.g. msgpack or JSON object).

Reply:
Based on our discussions in meetings, I've added a handshake for detecting the protocol version used by `mender-connect`. As a follow up we should add a test case using `mender-connect v1.0` and the old test case.

Reply:
Ok, I created https://tracker.mender.io/browse/MEN-4533.